File: scripts/sites.py
#!/usr/bin/env python3
"""입지 유형·태그·산단 외 기업 목록 — 규칙은 config/site_types.yml 에만 있고 이 모듈이 적용한다.
src/lib/sites.ts 가 같은 규칙을 같은 순서로 구현한다.

- site_type(complex, address) → 유형 이름 (수성알파시티 / 연구개발특구 / 지식산업센터 / 산업단지 / 개별입지)
- load_all_companies() → 팩토리온 기업 + scripts/data/extra_companies.csv(산단 외 기업) 를 한 목록으로. 각 행에 site_type, tags 채움
- python3 scripts/sites.py → 유형별·태그별 분포 출력
"""
from __future__ import annotations

import csv
import re
import sys
from collections import Counter
from pathlib import Path

import yaml

ROOT = Path(__file__).resolve().parent.parent
CONFIG = ROOT / "config" / "site_types.yml"
DATA = ROOT / "scripts" / "data"
COMPANIES = DATA / "dalseong_companies.csv"
EXTRA = DATA / "extra_companies.csv"
TAGS = DATA / "company_tags.csv"

_cfg: dict | None = None


def config() -> dict:
    global _cfg
    if _cfg is None:
        _cfg = yaml.safe_load(CONFIG.read_text(encoding="utf-8"))
    return _cfg


def site_types() -> list[dict]:
    return [{"key": t["key"], "name": t["name"]} for t in config()["types"]]


def site_type(complex_: str, address: str) -> str:
    cx = (complex_ or "").strip()
    addr = (address or "").replace(" ", "")
    for t in config()["types"]:
        if cx and cx in t.get("complexes", []):
            return t["name"]
        if any(p.replace(" ", "") in addr for p in t.get("address", [])):
            return t["name"]
        if t.get("rule") == "in_complex" and cx and cx not in ("개별입지", config()["outside_complex"]):
            return t["name"]
        if t.get("rule") == "default":
            return t["name"]
    return config()["types"][-1]["name"]


def tag_name(key: str) -> str:
    return config()["tags"].get(key, {}).get("name", key)


def is_startup(founded: str, as_of: str) -> bool:
    """founded(YYYY 또는 YYYY-MM…) 가 as_of 기준 7년 이내면 창업기업."""
    y = re.match(r"(\d{4})", founded or "")
    a = re.match(r"(\d{4})", as_of or "")
    return bool(y and a and int(a.group(1)) - int(y.group(1)) <= config()["startup_years"])


def load_tags() -> dict[str, set[str]]:
    out: dict[str, set[str]] = {}
    if TAGS.exists():
        for r in csv.DictReader(open(TAGS, encoding="utf-8")):
            if r.get("id") and r.get("tag"):
                out.setdefault(r["id"], set()).add(r["tag"])
    return out


def load_all_companies() -> list[dict]:
    """팩토리온 기업(모두) + 산단 외 기업. 열은 dalseong_companies.csv 와 같고 site_type, tags(집합), founded 가 추가된다."""
    rows = list(csv.DictReader(open(COMPANIES, encoding="utf-8")))
    for r in rows:
        r.setdefault("founded", "")
    if EXTRA.exists():
        for r in csv.DictReader(open(EXTRA, encoding="utf-8")):
            rows.append(r)
    tags = load_tags()
    for r in rows:
        r["site_type"] = site_type(r.get("complex", ""), r.get("address", ""))
        t = set(tags.get(r["id"], set()))
        for k in (r.get("tags") or "").split(";"):
            if k.strip():
                t.add(k.strip())
        if r.get("founded") and is_startup(r["founded"], r.get("as_of", "")):
            t.add("startup")
        r["tags"] = t
    return rows


def main() -> int:
    rows = load_all_companies()
    fo = sum(1 for r in rows if not r["id"].startswith("x"))
    print(f"기업 {len(rows):,}곳 (팩토리온 {fo:,} + 산단 외 목록 {len(rows) - fo:,})")
    print("\n입지 유형별")
    for name, n in Counter(r["site_type"] for r in rows).most_common():
        print(f"  {name:<10}{n:>8,}")
    tc = Counter(t for r in rows for t in r["tags"])
    print("\n태그별" + ("" if tc else " (없음 — scripts/data/extra/ 에 목록을 넣고 import_extra.py 실행)"))
    for k, n in tc.most_common():
        print(f"  {tag_name(k):<10}{n:>8,}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/lib/sites.ts
import fs from 'node:fs';
import yaml from 'js-yaml';

/** 입지 유형·태그. 규칙은 config/site_types.yml 에만 있다 (scripts/sites.py 와 같은 판정 순서). */
type SiteType = { key: string; name: string; complexes?: string[]; address?: string[]; buildings_file?: string; rule?: 'in_complex' | 'default' };
type Config = { version: string; types: SiteType[]; outside_complex: string; tags: Record<string, { name: string; desc: string }>; startup_years: number };
let _cfg: Config | null = null;
const cfg = () => (_cfg ??= yaml.load(fs.readFileSync('config/site_types.yml', 'utf-8')) as Config);

export const siteTypes = () => cfg().types.map(t => ({ key: t.key, name: t.name }));
export const siteKey = (name: string) => cfg().types.find(t => t.name === name)?.key ?? null;
export const siteName = (key: string) => cfg().types.find(t => t.key === key)?.name ?? null;
export const OUTSIDE = () => cfg().outside_complex;
export const tagName = (key: string) => cfg().tags[key]?.name ?? key;
export const tagDesc = (key: string) => cfg().tags[key]?.desc ?? '';
export const allTags = () => Object.entries(cfg().tags).map(([key, v]) => ({ key, ...v }));

const _buildings = new Map<string, RegExp[]>();
/** 건물 주소 패턴: '구군 + 도로명 건물번호'(공백 제거) 뒤에 숫자·'-'가 오지 않아야 한다 (scripts/sites.py 와 같은 규칙). */
function buildings(t: SiteType): RegExp[] {
  if (!t.buildings_file) return [];
  if (!_buildings.has(t.buildings_file)) {
    const pats: RegExp[] = [];
    if (fs.existsSync(t.buildings_file)) {
      const lines = fs.readFileSync(t.buildings_file, 'utf-8').replace(/^\uFEFF/, '').split(/\r?\n/).filter(Boolean);
      const head = lines[0].split(',');
      const di = head.indexOf('district'), ri = head.indexOf('road_address');
      for (const l of lines.slice(1)) {
        const c = l.split(',');
        const key = ((c[di] ?? '') + (c[ri] ?? '')).replace(/\s/g, '').replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        if (key) pats.push(new RegExp(key + '(?![\\d-])'));
      }
    }
    _buildings.set(t.buildings_file, pats);
  }
  return _buildings.get(t.buildings_file)!;
}
export function siteType(complex: string, address: string): string {
  const cx = (complex ?? '').trim();
  const addr = (address ?? '').replace(/\s/g, '');
  for (const t of cfg().types) {
    if (cx && (t.complexes ?? []).includes(cx)) return t.name;
    if ((t.address ?? []).some(p => addr.includes(p.replace(/\s/g, '')))) return t.name;
    if (buildings(t).some(b => b.test(addr))) return t.name;
    if (t.rule === 'in_complex' && cx && cx !== '개별입지' && cx !== cfg().outside_complex) return t.name;
    if (t.rule === 'default') return t.name;
  }
  return cfg().types.at(-1)!.name;
}
export function isStartup(founded: string, asOf: string): boolean {
  const y = /(\d{4})/.exec(founded ?? ''), a = /(\d{4})/.exec(asOf ?? '');
  return !!(y && a && Number(a[1]) - Number(y[1]) <= cfg().startup_years);
}
